File: backend/api/blocklist.py
"""FastAPI-Router der blocklist-Domaene (v2), prefix ``/api/blocklist``.

Aeusserer Ring: nimmt HTTP entgegen. Wie ``api/maintenance.py``/``api/report.py`` kennt
dieser Router WEDER ``application`` NOCH ``domain`` NOCH ``infrastructure`` (import-linter
Regel 4): alle Runner kommen als ``Callable`` bzw. schmale lokale ``Protocol``-Vertraege
per Dependency herein, verdrahtet im Composition Root (``app.py``). Die Projektion
Domaene->Wire macht der Composition-Root-Runner in ``app.py``, NICHT dieser Router.

Provider-Marker (``provide_*`` -> ``NotImplementedError``): ohne Verdrahtung bewusst ein
lauter Fehler (kein stiller Fallback, S3). Eigene pydantic ``*Out``/``*Body``-Modelle
(snake_case Wire). Vokabular-Hebung (group/fmt/strictness) passiert im Use-Case; ein
Fehlwert wirft dort ``ValueError`` (Application-``BlocklistError`` ist eine Unterklasse) ->
hier auf 422 gemappt. Unbekannte id bei PATCH/DELETE-naher Logik -> 404. 422/404 als
NACKTE Literale (Bestandsmuster ``api/monitoring.py``).
"""

from collections.abc import Callable
from typing import Annotated, Protocol

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

router = APIRouter(prefix="/api/blocklist", tags=["blocklist"])


# ── Wire-Modelle (eigene Form, KEIN application/domain-Typ) ────────────────────


class SourceOut(BaseModel):
    """Eine Quellen-Definition in Wire-Form (Projektion macht der Composition Root)."""

    id: str
    name: str
    group: str
    fmt: str
    origin: str
    url: str | None
    license: str
    attribution_required: bool
    enabled: bool
    last_fetched_ts: float | None
    status: str
    entry_count: int | None


class AddSourceBody(BaseModel):
    """POST /sources -- neue Nutzer-Quelle per URL. group/fmt roh (Hebung im Use-Case)."""

    name: str
    url: str
    group: str
    fmt: str


class AddSourceOut(BaseModel):
    """Ergebnis von POST /sources: vergebene id + freundlicher Lizenz-Hinweis (oder null)."""

    source_id: str
    license_hint: str | None


class UploadSourceBody(BaseModel):
    """POST /sources/upload -- hochgeladene Liste. ``content`` ist der rohe Listentext."""

    name: str
    group: str
    fmt: str
    content: str


class UploadSourceOut(BaseModel):
    """Ergebnis von POST /sources/upload: vergebene id + Zahl sofort geparster Eintraege."""

    source_id: str
    entry_count: int


class UpdateSourceBody(BaseModel):
    """PATCH /sources/{id} -- partielles Update; alle Felder optional (None = unveraendert)."""

    name: str | None = None
    url: str | None = None
    group: str | None = None
    fmt: str | None = None
    enabled: bool | None = None


class RefreshOut(BaseModel):
    """Ergebnis eines Lade-Laufs (POST /sources/{id}/refresh)."""

    source_id: str
    ok: bool
    entry_count: int | None
    error: str | None


class RefreshDueOut(BaseModel):
    """Ergebnis von POST /refresh-due: die Lade-Ergebnisse aller faelligen Quellen."""

    results: list[RefreshOut]


class HealthIssueOut(BaseModel):
    """Ein Gesundheits-Befund in Wire-Form (BROKEN-Quelle + optionaler Ersatzvorschlag)."""

    source_id: str
    name: str
    group: str
    suggested_replacement_id: str | None


class HealthOut(BaseModel):
    """Ergebnis von GET /health: die Liste der Gesundheits-Befunde."""

    issues: list[HealthIssueOut]


class SettingsOut(BaseModel):
    """Ergebnis von GET /settings: Strenge + Refresh-Intervall + Gruppen-Schalter."""

    strictness: str
    refresh_interval_days: int
    group_tracker_ads_enabled: bool
    group_threat_enabled: bool


class SettingsBody(BaseModel):
    """PUT /settings -- alle Felder optional (nur gesetzte werden geschrieben)."""

    strictness: str | None = None
    refresh_interval_days: int | None = None
    group_tracker_ads_enabled: bool | None = None
    group_threat_enabled: bool | None = None


class ContactInputBody(BaseModel):
    """Ein abzugleichender Aussenkontakt im Wire-Body von POST /match."""

    remote_ip: str
    hostname: str | None = None


class MatchBody(BaseModel):
    """POST /match -- Kontakte + optionale Strenge (fehlt sie, liest der Runner Settings)."""

    contacts: list[ContactInputBody]
    strictness: str | None = None


class MatchOut(BaseModel):
    """Ein Treffer in Wire-Form (Quelle + Wert, auf den getroffen wurde)."""

    source_id: str
    source_name: str
    group: str
    matched_on: str


class ContactMatchOut(BaseModel):
    """Das Abgleich-Ergebnis EINES Kontakts (auch leere ``matches`` zulaessig)."""

    remote_ip: str
    hostname: str | None
    matches: list[MatchOut]


class MatchResultsOut(BaseModel):
    """Ergebnis von POST /match: je Eingangs-Kontakt ein ``ContactMatchOut``."""

    results: list[ContactMatchOut]


# ── injizierte Composition-Root-Runner (Marker + schmale Vertraege) ────────────
# Provider-Marker: in app.py per dependency_overrides verdrahtet. Ohne Verdrahtung
# bewusst ein lauter Fehler (kein stiller Fallback, S3).


class AddSourceRunner(Protocol):
    """Schmaler Vertrag des injizierten Add-Runners (rohe Strings rein, Wire-Out raus).

    Die group/fmt-Hebung passiert im Use-Case; ein Fehlwert wirft ``ValueError``
    (Application-``BlocklistError`` ist Unterklasse) -> der Endpunkt mappt auf 422.
    """

    def __call__(self, body: AddSourceBody) -> AddSourceOut:
        """Legt die Quelle an und liefert id + Lizenz-Hinweis (Wire-Form)."""
        ...


class UploadSourceRunner(Protocol):
    """Schmaler Vertrag des injizierten Upload-Runners (parst sofort, Wire-Out raus)."""

    def __call__(self, body: UploadSourceBody) -> UploadSourceOut:
        """Importiert die hochgeladene Liste und liefert id + Eintragszahl."""
        ...


class UpdateSourceRunner(Protocol):
    """Schmaler Vertrag des injizierten Update-Runners (partielles Update).

    Unbekannte id wirft ``KeyError`` (der Composition-Root-Runner uebersetzt den
    Use-Case-Fehler in ``KeyError``) -> 404; ein Vokabular-Fehlwert wirft ``ValueError``
    -> 422.
    """

    def __call__(self, source_id: str, body: UpdateSourceBody) -> None:
        """Aktualisiert die Quelle partiell (nur gesetzte Felder)."""
        ...


class DeleteSourceRunner(Protocol):
    """Schmaler Vertrag des injizierten Delete-Runners (idempotent)."""

    def __call__(self, source_id: str) -> None:
        """Loescht die Quelle samt Eintraegen (idempotent)."""
        ...


class RefreshSourceRunner(Protocol):
    """Schmaler Vertrag des injizierten Einzel-Refresh-Runners.

    Unbekannte id / Upload-ohne-url wirft ``KeyError`` -> 404; ein Download-/Parse-Fehler
    ist KEIN HTTP-Fehler -> er kommt als ``RefreshOut(ok=False, ...)`` zurueck.
    """

    def __call__(self, source_id: str) -> RefreshOut:
        """Laedt EINE Quelle und liefert das Lade-Ergebnis (Wire-Form)."""
        ...


class MatchRunner(Protocol):
    """Schmaler Vertrag des injizierten Match-Runners (Strenge aus Body ODER Settings).

    Eine im Body angegebene, aber unbekannte Strenge wirft ``ValueError`` -> 422.
    """

    def __call__(self, body: MatchBody) -> MatchResultsOut:
        """Gleicht die Kontakte ab und liefert je Kontakt das Treffer-Ergebnis."""
        ...


class SettingsWriteRunner(Protocol):
    """Schmaler Vertrag des injizierten Settings-Schreib-Runners.

    Eine unbekannte Strenge im Body wirft ``ValueError`` -> 422.
    """

    def __call__(self, body: SettingsBody) -> None:
        """Schreibt die gesetzten Settings-Felder (Strenge wird validiert)."""
        ...


# Lese-Runner als reine Callables (liefern die fertige Wire-Sicht).
type SourceListProvider = Callable[[], Callable[[], list[SourceOut]]]
type HealthProvider = Callable[[], Callable[[], HealthOut]]
type SettingsReadProvider = Callable[[], Callable[[], SettingsOut]]
type RefreshDueProvider = Callable[[], Callable[[], RefreshDueOut]]
type ResetDefaultsProvider = Callable[[], Callable[[], None]]


def provide_list_sources() -> Callable[[], list[SourceOut]]:
    raise NotImplementedError("list_sources wird in app.py verdrahtet")


def provide_add_source() -> AddSourceRunner:
    raise NotImplementedError("AddSourceRunner wird in app.py verdrahtet")


def provide_upload_source() -> UploadSourceRunner:
    raise NotImplementedError("UploadSourceRunner wird in app.py verdrahtet")


def provide_update_source() -> UpdateSourceRunner:
    raise NotImplementedError("UpdateSourceRunner wird in app.py verdrahtet")


def provide_delete_source() -> DeleteSourceRunner:
    raise NotImplementedError("DeleteSourceRunner wird in app.py verdrahtet")


def provide_refresh_source() -> RefreshSourceRunner:
    raise NotImplementedError("RefreshSourceRunner wird in app.py verdrahtet")


def provide_refresh_due() -> Callable[[], RefreshDueOut]:
    raise NotImplementedError("refresh_due wird in app.py verdrahtet")


def provide_reset_defaults() -> Callable[[], None]:
    raise NotImplementedError("reset_defaults wird in app.py verdrahtet")


def provide_health() -> Callable[[], HealthOut]:
    raise NotImplementedError("health wird in app.py verdrahtet")


def provide_read_settings() -> Callable[[], SettingsOut]:
    raise NotImplementedError("read_settings wird in app.py verdrahtet")


def provide_write_settings() -> SettingsWriteRunner:
    raise NotImplementedError("SettingsWriteRunner wird in app.py verdrahtet")


def provide_match() -> MatchRunner:
    raise NotImplementedError("MatchRunner wird in app.py verdrahtet")


# ── Routen ───────────────────────────────────────────────────────────────────


@router.get("/sources")
def list_sources(
    runner: Annotated[Callable[[], list[SourceOut]], Depends(provide_list_sources)],
) -> list[SourceOut]:
    """Liefert alle Quellen-Definitionen (Wire-Form; Projektion im Composition Root)."""
    return runner()


@router.post("/sources")
def add_source(
    body: AddSourceBody,
    runner: Annotated[AddSourceRunner, Depends(provide_add_source)],
) -> AddSourceOut:
    """Legt eine Nutzer-Quelle per URL an. group/fmt ungueltig -> 422 (Use-Case wirft)."""
    try:
        return runner(body)
    except ValueError as exc:
        raise HTTPException(422, detail=str(exc)) from exc


@router.post("/sources/upload")
def upload_source(
    body: UploadSourceBody,
    runner: Annotated[UploadSourceRunner, Depends(provide_upload_source)],
) -> UploadSourceOut:
    """Importiert eine hochgeladene Liste (parst sofort). group/fmt ungueltig -> 422."""
    try:
        return runner(body)
    except ValueError as exc:
        raise HTTPException(422, detail=str(exc)) from exc


@router.patch("/sources/{source_id}")
def update_source(
    source_id: str,
    body: UpdateSourceBody,
    runner: Annotated[UpdateSourceRunner, Depends(provide_update_source)],
) -> dict[str, bool]:
    """Partielles Update. Unbekannte id -> 404; Vokabular-Fehlwert -> 422."""
    try:
        runner(source_id, body)
    except KeyError as exc:
        raise HTTPException(404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(422, detail=str(exc)) from exc
    return {"ok": True}


@router.delete("/sources/{source_id}")
def delete_source(
    source_id: str,
    runner: Annotated[DeleteSourceRunner, Depends(provide_delete_source)],
) -> dict[str, bool]:
    """Loescht eine Quelle samt Eintraegen (idempotent)."""
    runner(source_id)
    return {"ok": True}


@router.post("/sources/{source_id}/refresh")
def refresh_source(
    source_id: str,
    runner: Annotated[RefreshSourceRunner, Depends(provide_refresh_source)],
) -> RefreshOut:
    """Laedt EINE Quelle. Unbekannte id / Upload-ohne-url -> 404.

    Ein Download-/Parse-Fehler ist KEIN HTTP-Fehler: er kommt als
    ``RefreshOut(ok=False, error=...)`` zurueck (status BROKEN, der Health-Check
    verarbeitet ihn).
    """
    try:
        return runner(source_id)
    except KeyError as exc:
        raise HTTPException(404, detail=str(exc)) from exc


@router.post("/refresh-due")
def refresh_due(
    runner: Annotated[Callable[[], RefreshDueOut], Depends(provide_refresh_due)],
) -> RefreshDueOut:
    """Manueller "alle faelligen jetzt": refresht die faelligen Quellen (Intervall aus Settings)."""
    return runner()


@router.post("/reset-defaults")
def reset_defaults(
    runner: Annotated[Callable[[], None], Depends(provide_reset_defaults)],
) -> dict[str, bool]:
    """Werkszustand der Listen: leert + legt alle Werksquellen neu an."""
    runner()
    return {"ok": True}


@router.get("/health")
def health(
    runner: Annotated[Callable[[], HealthOut], Depends(provide_health)],
) -> HealthOut:
    """Liefert je BROKEN-Quelle einen Befund + optionalen Ersatzvorschlag."""
    return runner()


@router.get("/settings")
def read_settings(
    runner: Annotated[Callable[[], SettingsOut], Depends(provide_read_settings)],
) -> SettingsOut:
    """Liefert Strenge + Refresh-Intervall + Gruppen-Schalter (Defaults bei leerer DB)."""
    return runner()


@router.put("/settings")
def write_settings(
    body: SettingsBody,
    runner: Annotated[SettingsWriteRunner, Depends(provide_write_settings)],
) -> dict[str, bool]:
    """Schreibt die gesetzten Settings-Felder. Strenge ungueltig -> 422."""
    try:
        runner(body)
    except ValueError as exc:
        raise HTTPException(422, detail=str(exc)) from exc
    return {"ok": True}


@router.post("/match")
def match(
    body: MatchBody,
    runner: Annotated[MatchRunner, Depends(provide_match)],
) -> MatchResultsOut:
    """Gleicht Aussenkontakte gegen die aktiven Listen ab. Strenge ungueltig -> 422.

    Fehlt ``strictness`` im Body, liest der Composition-Root-Runner sie aus den Settings;
    die Gruppen-Feinschalter aus den Settings fliessen ebenfalls ein.
    """
    try:
        return runner(body)
    except ValueError as exc:
        raise HTTPException(422, detail=str(exc)) from exc

File: frontend/src/api/blocklist.js
// Blocklist-API-Client (CERNIS PRO 2.0)
//
// Duenne, reine Wrapper um die Blocklist-Endpunkte (Verwaltung extern gepflegter
// Listen, Abgleich von Aussenkontakten). Stil bewusst wie api/report.js /
// api/maintenance.js: kleine reine Mapper, snake_case <-> camelCase, KEIN erfundener
// Fallback (null bleibt ehrlich null). Pfade bleiben relativ; NIE einen Host
// hartkodieren.
//
// client.js hat KEIN apiPatch und KEIN apiDelete. Beide werden hier lokal per fetch
// gebaut — mit derselben Fehler-Form wie client.js (ApiError mit status/message),
// die wir aus client.js importieren. PUT/POST/GET laufen ueber die vorhandenen
// apiGet/apiPost/apiPut.
//
// Wire-Form (verifiziert gegen backend/api/blocklist.py, NICHT aendern):
//   GET    /api/blocklist/sources               -> [SourceOut]
//   POST   /api/blocklist/sources               -> { source_id, license_hint }
//   POST   /api/blocklist/sources/upload        -> { source_id, entry_count }
//   PATCH  /api/blocklist/sources/{id}          -> { ok }
//   DELETE /api/blocklist/sources/{id}          -> { ok }
//   POST   /api/blocklist/sources/{id}/refresh  -> { source_id, ok, entry_count, error }
//   POST   /api/blocklist/refresh-due           -> { results: [RefreshOut] }
//   POST   /api/blocklist/reset-defaults        -> { ok }
//   GET    /api/blocklist/health                -> { issues: [HealthIssueOut] }
//   GET    /api/blocklist/settings              -> SettingsOut
//   PUT    /api/blocklist/settings              -> { ok }
//   POST   /api/blocklist/match                 -> { results: [ContactMatchOut] }

import { ApiError, apiGet, apiPost, apiPut } from "./client.js";

// ── Lokale PATCH/DELETE-Helfer (client.js hat sie nicht) ───────────────────────
// Bewusst exakt analog zu apiPut/apiPost in client.js: gleiche Header, gleiche
// ApiError-Form (status = HTTP-Code oder null bei Netzfehler).

// PATCH auf einen relativen API-Pfad mit JSON-Body. Gleiche Fehler-Form wie apiPut.
async function apiPatch(path, body) {
  let response;
  try {
    response = await fetch(path, {
      method: "PATCH",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
  } catch (ursache) {
    // Netzfehler (Server nicht erreichbar, DNS, Abbruch o. Ä.): kein HTTP-Status.
    throw new ApiError(ursache?.message ?? "Netzwerkfehler", null);
  }

  if (!response.ok) {
    throw new ApiError(`Unerwarteter HTTP-Status ${response.status}`, response.status);
  }

  return response.json();
}

// DELETE auf einen relativen API-Pfad (kein Body). Gleiche Fehler-Form wie apiPut.
async function apiDelete(path) {
  let response;
  try {
    response = await fetch(path, {
      method: "DELETE",
      headers: { Accept: "application/json" },
    });
  } catch (ursache) {
    // Netzfehler (Server nicht erreichbar, DNS, Abbruch o. Ä.): kein HTTP-Status.
    throw new ApiError(ursache?.message ?? "Netzwerkfehler", null);
  }

  if (!response.ok) {
    throw new ApiError(`Unerwarteter HTTP-Status ${response.status}`, response.status);
  }

  return response.json();
}

// ── Mapper (snake_case -> camelCase) ───────────────────────────────────────────

// Eine Quellen-Definition -> View-Struktur. Alle Felder roh durchgereicht
// (group/fmt/origin/status tragen das Backend-Vokabular unveraendert; null bleibt
// ehrlich null bei url/lastFetchedTs/entryCount).
function mappeSource(s) {
  return {
    id: s.id,
    name: s.name,
    group: s.group,
    fmt: s.fmt,
    origin: s.origin,
    url: s.url ?? null,
    license: s.license,
    attributionRequired: s.attribution_required,
    enabled: s.enabled,
    lastFetchedTs: s.last_fetched_ts ?? null,
    status: s.status,
    entryCount: s.entry_count ?? null,
  };
}

// Ein Lade-Ergebnis (Einzel-Refresh / refresh-due) -> View-Struktur.
function mappeRefresh(r) {
  return {
    sourceId: r.source_id,
    ok: r.ok,
    entryCount: r.entry_count ?? null,
    error: r.error ?? null,
  };
}

// Ein Gesundheits-Befund -> View-Struktur.
function mappeHealthIssue(i) {
  return {
    sourceId: i.source_id,
    name: i.name,
    group: i.group,
    suggestedReplacementId: i.suggested_replacement_id ?? null,
  };
}

// Ein Treffer eines Kontakts -> View-Struktur.
function mappeMatch(m) {
  return {
    sourceId: m.source_id,
    sourceName: m.source_name,
    group: m.group,
    matchedOn: m.matched_on,
  };
}

// Das Abgleich-Ergebnis EINES Kontakts -> View-Struktur.
function mappeContactMatch(c) {
  return {
    remoteIp: c.remote_ip,
    hostname: c.hostname ?? null,
    matches: (c.matches ?? []).map(mappeMatch),
  };
}

// ── Endpunkt-Wrapper ───────────────────────────────────────────────────────────

// GET /sources -> alle Quellen-Definitionen (Wire-Form, hier camelCase gemappt).
export async function fetchSources() {
  const backend = await apiGet("/api/blocklist/sources");
  return (backend ?? []).map(mappeSource);
}

// POST /sources -> legt eine Nutzer-Quelle per URL an. group/fmt sind rohe Wire-
// Strings (Hebung im Backend; 422 bei ungueltigem Wert). Antwort: vergebene id +
// freundlicher Lizenz-Hinweis (oder null — kein erfundener Hinweis).
export async function addSource({ name, url, group, fmt }) {
  const backend = await apiPost("/api/blocklist/sources", { name, url, group, fmt });
  return {
    sourceId: backend.source_id,
    licenseHint: backend.license_hint ?? null,
  };
}

// POST /sources/upload -> importiert eine hochgeladene Liste. content ist der rohe
// Listentext. Antwort: vergebene id + Zahl sofort geparster Eintraege.
export async function uploadSource({ name, group, fmt, content }) {
  const backend = await apiPost("/api/blocklist/sources/upload", {
    name,
    group,
    fmt,
    content,
  });
  return {
    sourceId: backend.source_id,
    entryCount: backend.entry_count,
  };
}

// PATCH /sources/{id} -> partielles Update (nur gesetzte Felder). patch ist bereits
// in Wire-Form (camelCase deckt sich hier mit snake_case: name/url/group/fmt/enabled).
export async function updateSource(id, patch) {
  return apiPatch(`/api/blocklist/sources/${encodeURIComponent(id)}`, patch);
}

// DELETE /sources/{id} -> loescht eine Quelle samt Eintraegen (idempotent).
export async function deleteSource(id) {
  return apiDelete(`/api/blocklist/sources/${encodeURIComponent(id)}`);
}

// POST /sources/{id}/refresh -> laedt EINE Quelle. Ein Download-/Parse-Fehler ist
// KEIN HTTP-Fehler: er kommt als { ok:false, error } zurueck.
export async function refreshSource(id) {
  const backend = await apiPost(
    `/api/blocklist/sources/${encodeURIComponent(id)}/refresh`,
  );
  return mappeRefresh(backend);
}

// POST /refresh-due -> refresht alle faelligen Quellen (Intervall aus Settings).
// Antwort: je Quelle ein Lade-Ergebnis.
export async function refreshDue() {
  const backend = await apiPost("/api/blocklist/refresh-due");
  return { results: (backend?.results ?? []).map(mappeRefresh) };
}

// POST /reset-defaults -> Werkszustand der Listen (leeren + Werksquellen neu anlegen).
export async function resetDefaults() {
  return apiPost("/api/blocklist/reset-defaults");
}

// GET /health -> je BROKEN-Quelle ein Befund + optionaler Ersatzvorschlag.
export async function fetchHealth() {
  const backend = await apiGet("/api/blocklist/health");
  return { issues: (backend?.issues ?? []).map(mappeHealthIssue) };
}

// GET /settings -> Strenge + Refresh-Intervall + Gruppen-Feinschalter (camelCase).
export async function fetchSettings() {
  const backend = await apiGet("/api/blocklist/settings");
  return {
    strictness: backend.strictness,
    refreshIntervalDays: backend.refresh_interval_days,
    groupTrackerAdsEnabled: backend.group_tracker_ads_enabled,
    groupThreatEnabled: backend.group_threat_enabled,
  };
}

// PUT /settings -> schreibt die gesetzten Settings-Felder. patch ist camelCase und
// wird hier auf snake_case abgebildet; nur definierte Felder werden gesendet
// (undefined faellt raus -> Backend laesst sie unveraendert). 422 bei ungueltiger
// Strenge (kein stiller Fallback).
export async function writeSettings(patch) {
  const body = {};
  if (patch.strictness !== undefined) {
    body.strictness = patch.strictness;
  }
  if (patch.refreshIntervalDays !== undefined) {
    body.refresh_interval_days = patch.refreshIntervalDays;
  }
  if (patch.groupTrackerAdsEnabled !== undefined) {
    body.group_tracker_ads_enabled = patch.groupTrackerAdsEnabled;
  }
  if (patch.groupThreatEnabled !== undefined) {
    body.group_threat_enabled = patch.groupThreatEnabled;
  }
  return apiPut("/api/blocklist/settings", body);
}

// POST /match -> gleicht Aussenkontakte gegen die aktiven Listen ab. contacts ist ein
// Array { remoteIp, hostname? }; strictness optional (fehlt sie, liest das Backend die
// Settings). Wird von CC-4 (Block-Ansicht) genutzt — hier schon mit-exportiert.
export async function matchContacts(contacts, strictness) {
  const body = {
    contacts: (contacts ?? []).map((c) => ({
      remote_ip: c.remoteIp,
      hostname: c.hostname ?? null,
    })),
  };
  if (strictness !== undefined && strictness !== null) {
    body.strictness = strictness;
  }
  const backend = await apiPost("/api/blocklist/match", body);
  return { results: (backend?.results ?? []).map(mappeContactMatch) };
}

export default {
  fetchSources,
  addSource,
  uploadSource,
  updateSource,
  deleteSource,
  refreshSource,
  refreshDue,
  resetDefaults,
  fetchHealth,
  fetchSettings,
  writeSettings,
  matchContacts,
};
